contracts/RewardsAggregator.sol

```diff
@@ -0,0 +1,262 @@
+    using SafeERC20 for IERC20;
+    using MerkleProof for bytes32[];
+
+    address public immutable override token;
+    address public rewardsHolder;
+
+    bytes32 public override merkleRoot;
+    mapping(address => uint256) internal cumulativeClaimed;
+
+    // TODO: Generalize to an array of IApplication in the future.
+    // For the moment, it will only be used for TACo app.
+    IApplication public immutable application;
+
+    ICumulativeMerkleDrop public immutable oldCumulativeMerkleDrop;
+
+    struct MerkleClaim {
+        address stakingProvider;
+        address beneficiary;
+        uint256 amount;
+        bytes32[] proof;
+    }
+
+    constructor(
+        address token_,
+        IApplication application_,
+        ICumulativeMerkleDrop _oldCumulativeMerkleDrop,
+        address rewardsHolder_,
+        address newOwner
+    ) {
+        require(IERC20(token_).totalSupply() > 0, "Token contract must be set");
+        require(
+            rewardsHolder_ != address(0),
+            "Rewards Holder must be an address"
+        );
+        require(
+            address(application_) != address(0),
+            "Application must be an address"
+        );
+        require(
+            token_ == _oldCumulativeMerkleDrop.token(),
+            "Incompatible old Merkle Distribution contract"
+        );
+
+        transferOwnership(newOwner);
+        token = token_;
+        application = application_;
+        rewardsHolder = rewardsHolder_;
+        oldCumulativeMerkleDrop = _oldCumulativeMerkleDrop;
+    }
+
+    function setMerkleRoot(bytes32 merkleRoot_) external override onlyOwner {
+        emit MerkleRootUpdated(merkleRoot, merkleRoot_);
+        merkleRoot = merkleRoot_;
+    }
+
+    function setRewardsHolder(address rewardsHolder_) external onlyOwner {
+        require(
+            rewardsHolder_ != address(0),
+            "Rewards Holder must be an address"
+        );
+        emit RewardsHolderUpdated(rewardsHolder, rewardsHolder_);
+        rewardsHolder = rewardsHolder_;
+    }
+
+    /**
+     * @notice Returns the amount of rewards that a given stake has already
+     *         claimed from the Merkle distribution, including the old Merkle
+     *         distribution contract. The returned amount does not include the
+     *         claimed rewards generated by the applications.
+     */
+    function cumulativeMerkleClaimed(
+        address stakingProvider
+    ) public view returns (uint256) {
+        uint256 newAmount = cumulativeClaimed[stakingProvider];
+        if (newAmount > 0) {
+            return newAmount;
+        } else {
+            return oldCumulativeMerkleDrop.cumulativeClaimed(stakingProvider);
+        }
+    }
+
+    /**
+     * @notice Claim the rewards that have been generated by the Merkle
+     *         distribution mechanism.
+     */
+    function claimMerkle(
+        address stakingProvider,
+        address beneficiary,
```

Comment:
Just to confirm for my benefit - the verification of the merkle proof is what allows `beneficiary` to be passed in to this function, as opposed to obtaining the `beneficiary` value from the contract ourselves?

Reply:
Correct. This contract is actually unaware of the staking contract.